pkg/operation/botanist/component/extensions/dns/dnsprovider.go

```diff
@@ -135,24 +150,39 @@ func (p *provider) Wait(ctx context.Context) error {
 		interval              = 5 * time.Second
 		severeThreshold       = 15 * time.Second
 		timeout               = 2 * time.Minute
+
+		annotationHealthFunc health.Func
 	)
 
-	if err := p.waiter.UntilTimeout(ctx, interval, timeout, func(ctx context.Context) (done bool, err error) {
-		obj := &dnsv1alpha1.DNSProvider{}
-		if err := p.client.Get(
-			ctx,
-			client.ObjectKey{Name: p.values.Name, Namespace: p.namespace},
-			obj,
-		); err != nil {
+	// wait until we see the timestamp annotation, that we set earlier in Deploy, to prevent falsely returning from Wait
+	// in case of stale cache reads
+	if expectedTimestamp, ok := p.dnsProvider.Annotations[v1beta1constants.GardenerTimestamp]; ok {
+		annotationHealthFunc = health.ObjectHasAnnotationWithValue(v1beta1constants.GardenerTimestamp, expectedTimestamp)
+	}
+
+	// TODO: switch to extensions.WaitUntilObjectReadyWithHealthFunction?
+	if err := p.waiter.UntilTimeout(ctx, interval, timeout, func(ctx context.Context) (bool, error) {
+		if err := p.client.Get(ctx, client.ObjectKeyFromObject(p.dnsProvider), p.dnsProvider); err != nil {
+			if apierrors.IsNotFound(err) {
+				return retry.MinorError(err)
+			}
 			return retry.SevereError(err)
 		}
 
-		if obj.Status.State == dnsv1alpha1.STATE_READY {
+		if annotationHealthFunc != nil {
+			if err := annotationHealthFunc(p.dnsProvider); err != nil {
+				return retry.MinorError(err)
+			}
+		}
+
+		// TODO: why are we not checking generation == observedGeneration here?
```

Comment:
Do you know, why we don't check `generation == observedGeneration` here for the `DNSProvider`, but we do it for the `DNSEntry`? https://github.com/gardener/gardener/blob/f995815edb908841886b546bd726ad114bf39c1a/pkg/operation/botanist/component/extensions/dns/dnsentry.go#L123

Reply:
No, I guess that this wasn't done on purpose was rather just forgotten. I don't see a problem with adding it.